### ISSUE TITLE
Image preprocessing in the server

### ISSUE THREAD

**Issue by R-Miner** [Rahza/.NET-Tensorflow-Classification-Client#1]
Can we do the image preprocessing in the server side rather than doing in the client side (I meant converting the image to tensor)?

**Comment by Rahza** [comment]
There is no preprocessing being done on the client side. The binary data is written to a string tensor. The image is then decoded and preprocessed on the server.

I don't know enough about grpc/protobuf to tell you if there is another way. You could always just write a REST wrapper service in Python so that you don't have to use grpc/protobuf on client side.

**Comment by R-Miner** [comment]
Thankyou for clarifying. 

When I try to generate a tensor, I get an exception saying "input size does not match signature" what does that mean? Do I need to define the string tensor according to signature? How?

**Comment by R-Miner** [comment]
inputs {
  key: "input"
  value {
    name: "input_1:0"
    dtype: DT_FLOAT
    tensor_shape {
      dim {
        size: -1
      }
      dim {
        size: -1
      }
      dim {
        size: -1
      }
      dim {
        size: 3
      }
    }
  }
}
outputs {
  key: "prob"
  value {
    name: "dense_2/Softmax:0"
    dtype: DT_FLOAT
    tensor_shape {
      dim {
        size: -1
      }
      dim {
        size: 2
      }
    }
  }
}

This is how my signature is.

**Comment by Rahza** [comment]
This is exactly what I don't know either. You can use the client as it is in this repo like this (for example):

```
Image img = Image.FromFile(@"C:\img.jpg");

byte[] arr;
using (MemoryStream ms = new MemoryStream())
{
    img.Save(ms, System.Drawing.Imaging.ImageFormat.Jpeg);
    arr =  ms.ToArray();
}

string host = "127.0.0.1:8500";
var predictionClient = new PredictionClient(Configuration.PredictionServer.Value);
var predictionResult = predictionClient.Predict(arr);
```

It only works with a single image at a time, I couldn't get it to work with multiple images yet. Your signature looks like it only accepts the float array as an input. I prepended a `tf.placeholder(tf.string, shape=[None, 1]` to my model and added operations for decoding/preprocessing.

**Comment by R-Miner** [comment]
I am getting the following error while defining the input tensor.

Grpc.Core.RpcException' occurred in mscorlib.dll
Status(StatusCode=InvalidArgument, Detail="input and filter must have the same depth: 1 vs 3
	 [[Node: conv2d_1/convolution = Conv2D[T=DT_FLOAT, data_format="NHWC", dilations=[1, 1, 1, 1], padding="VALID", strides=[1, 2, 2, 1], use_cudnn_on_gpu=true, _device="/job:localhost/replica:0/task:0/device:CPU:0"](_arg_input_1_0_0, conv2d_1/kernel/read)]]")

Found solutions at the python end but not for c# client. Any help is appreciated. thanks in advance.

**Comment by R-Miner** [comment]
Do you have any pointers to documentation on prepending placeholder to the model? Could you share the string placeholder preprocessing to required type?

Did you use decode_raw to convert the strong tensor back to float or required type?

**Comment by Rahza** [comment]
Well, I'm assuming you have some kind of `predict(img)` function. I personally have solved it like this (again, inelegantly - note that I'm exclusively using JPEG but you should be able to adapt the code to support multiple image formats):

```
inputs = tf.placeholder(tf.string, shape=[None, 1])
images = tf.map_fn(lambda x: tf.image.decode_jpeg(tf.squeeze(x), channels=3), inputs, dtype=tf.uint8, back_prop=False, parallel_iterations=16)
prediction = predict(images)
```

**Comment by R-Miner** [comment]
Thankyou for your quick response. Its of much help. I was stumbled many times in the process of tensorflow serving prediction. This will try and let you know how it went.

**Comment by R-Miner** [comment]
HI  Rahza, Do you export the model at any point of time??

**Comment by Rahza** [comment]
Yes. Basically, I was training the model, saved it, then constructed the "new" graph with the prepended string placeholder, loaded the saved model and then exported this new graph. I essentially followed this tutorial: https://www.tensorflow.org/serving/serving_basic#train_and_export_tensorflow_model

**Comment by R-Miner** [comment]
Yes I have a client -server tensorflow serving working perfectly in ubuntu machine. I am trying to create a c# client which now sends a string tensor. My server by default accepts a float type and tried to built a float tensor and pass it from client which failed. Aftet inputs from you,  I now sent a string tensor. Now I have the prepend layer of the model with the conversion from string to float where I go wrong. Couldn't figure out where I am going wrong. 

Do you have a reference to  how to prepend a layer to exisiting model?

**Comment by R-Miner** [comment]
The following is the way I tried to prepend my simple Keras model. 

# Regular model, trained as usual
parent_model = 'parent_model.h5'

string_input = tf.placeholder(tf.string,shape=(None,1))
images = tf.map_fn(lambda x: tf.image.decode_jpeg(tf.squeeze(x), channels=3),
                   string_input, dtype=tf.uint8)
image = tf.image.convert_image_dtype(images, dtype=tf.float32)
image = tf.expand_dims(image, 0)
print(image.shape)

# Preprocessing model
preproc_model = Sequential()
preproc_model.add((image)) **# Is this the way to add the layer to the model? This is where I get the error**


# Prepend the preprocessing model to parent model    
full_model = Model(inputs=[preproc_model.input],
              outputs=[parent_model(preproc_model.output)])

# Save the complete model to disk
full_model.save('full_model.h5')

**# Is this the way to add the layer to the model? That is where I get the error**

**Comment by Rahza** [comment]
I'm using native tensorflow code, I'm afraid I can't help you with Keras. But you shouldn't need a "preprocessing model". I can only give you another TensorFlow example. You have some kind of inference function:

```
def inference(images):
  conv1 = tf.layers.conv2d(inputs=images, filters=32, kernel_size=[5, 5], padding="same", activation=tf.nn.relu)
  ....
  logits = tf.layers.dense(inputs=conv5, axis=1)
  return logits
```

That you can train on images. Once done, you export the model. Then you can define a function:

```
def export():
  inputs = tf.placeholder(tf.string, shape=[None, 1])
  images = tf.map_fn(...)
  prediction = inference(images)
  # Restore saved model (will load weights for the previously trained part of the graph)
  # Export "new" model (which now accepts a string tensor as input)
```

**Comment by R-Miner** [comment]
Rahza

So you have the same NN(covolutional layer, maxpooling, dense layer etc...) like the saved model in the function 'inference' and do prediction with the new preprocessed image, followed by restoring the saved model and exporting it. 

And in prediction = inference(images) , what is the data type of variable prediction? Where do you use the value in this prediction variable.
Please correct me if I am wrong. Thanks  lot for your time. Greatly appreciate it.

**Comment by Rahza** [comment]
Yes, in fact, [inference()](url) is the exact same function. You could either copy&paste it to your export program or import the Python module that contains the `inference()` function and use the exact same one.

In my case `prediction` contains whatever tensor the model outputs. For example a tensor with shape `[batch_size, num_labels]` which, for each image, contains the score for each label.

**Comment by R-Miner** [comment]
I  followed all these steps but getting some errors. Will try again. Thankyou! Do you have any suggestions on good books on tesnorflow or tensorflow serving?

**Comment by Rahza** [comment]
Which errors do you get? Unfortunately I don't have any specific recommendations, but you'll find a lot of great resources on the internet. It's a lot of trial and error.

**Comment by R-Miner** [comment]
"slice index 0 of dimension 0 out of bounds" this is the error I get.

How did you create the image string input? This is how I did it.....


At client side:
    img = cv2.imread('picture/abc.png')
    img_str = cv2.imencode('.png', img)[1].tostring()
    print(type(img_str)) 

img_str is a <bytes> type
This is the input to the server.

At export:
string_input = tf.placeholder(tf.string,shape=(None,1))
print(string_input.shape)
images = tf.map_fn(lambda x: tf.image.decode_png(tf.squeeze(x), channels=3),
                   string_input, dtype=tf.uint8)
images=tf.reshape(images,[256,256,3])
images = tf.image.convert_image_dtype(images, dtype=tf.float32)
images = tf.expand_dims(images, 0)
print(images.shape)

Do you find any mistake in this?

**Comment by R-Miner** [comment]
Thankyou for all the help! I finally got to input a string image to a keras model in tensorflow serving. Altered the map_fn and it worked.
